deploy-edge.js

```diff
@@ -0,0 +1,170 @@
+      FunctionName: argv.functionName,
+      Handler: argv.handler,
+      Environment: {
+        Variables: argv.lambdaEnvInject ? {} : envVars
+      }
+    });
+
+    console.log(`Waiting for Lambda function '${argv.functionName}' update to be completed`);
+    await lambdaWaitFor('functionUpdated', {
+      FunctionName: argv.functionName
+    });
+
+    console.log(`Publishing Lambda function '${argv.functionName}'`);
+    const lambdaPublishResult = await lambdaPublishVersion({
+      FunctionName: argv.functionName
+    });
+
+    console.log(`Waiting for Lambda function '${argv.functionName}' update to be completed`);
+    await lambdaWaitFor('functionUpdated', {
+      FunctionName: argv.functionName
+    });
+
+    console.log(`Successfully published Lambda function '${argv.functionName}'`);
+    console.log(`  * Function ARN: ${lambdaPublishResult.FunctionArn}`);
+    console.log(`  * Version: ${lambdaPublishResult.Version}`);
+
+    console.log(`Fetching configuration for CloudFront distribution '${argv.cfDistributionId}'`);
+    const currentDistributionConfig = await cfGetDistributionConfig({
+      Id: argv.cfDistributionId
+    });
+
+    let hasUpdates = false;
+
+    const nextDistributionConfig = JSON.parse(JSON.stringify(currentDistributionConfig));
+    delete nextDistributionConfig.ETag;
+    nextDistributionConfig.Id = argv.cfDistributionId;
+    nextDistributionConfig.IfMatch = currentDistributionConfig.ETag;
+    for (const cacheBehavior of nextDistributionConfig.DistributionConfig.CacheBehaviors.Items) {
+      for (const association of cacheBehavior.LambdaFunctionAssociations.Items) {
+        if (association.LambdaFunctionARN.startsWith(lambdaUpdateResult.FunctionArn)) {
+          console.log('Updating Lambda Function ARN');
+          console.log(`  * From: ${association.LambdaFunctionARN}`);
+          console.log(`  * To: ${lambdaPublishResult.FunctionArn}`);
+          association.LambdaFunctionARN = lambdaPublishResult.FunctionArn;
+          hasUpdates = true;
+        }
+      }
+    }
+
+    if (!hasUpdates) {
+      throw new Error('CloudFront configuration has no function to update');
+    }
+
+    console.log(`Updating CloudFront distribution '${argv.cfDistributionId}'`);
+    const { ETag: etagAfterUpdate } = await cfUdateDistribution(nextDistributionConfig);
+    if (etagAfterUpdate !== currentDistributionConfig.ETag) {
+      console.log(`Successfully updated CloudFront distribution '${argv.cfDistributionId}':`);
+      console.log(`  * Old etag: ${currentDistributionConfig.ETag}`);
+      console.log(`  * New etag: ${etagAfterUpdate}`);
+    } else {
+      throw new Error('CloudFront distribution update was not successful, the etag is still the same');
+    }
+
+    if (argv.wait) {
+      console.log('Waiting for distribution deployment...');
+      await cfWaitFor('distributionDeployed', {
+        Id: argv.cfDistributionId
+      });
+    }
+
+    console.log('DONE!');
+  }
```

Comment:
I also won't pretend I fully understand or can validate the logic here and alot is being parsed and invoked.
A small presentation of the whole integration afterwards will be useful and appreciated.

Reply:
Sure